internal/receiver/discoveryreceiver/correlation.go

```diff
@@ -0,0 +1,217 @@
+// Copyright Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discoveryreceiver
+
+import (
+	"sync"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+	"go.opentelemetry.io/collector/config"
+	"go.uber.org/zap"
+
+	"github.com/signalfx/splunk-otel-collector/internal/receiver/discoveryreceiver/statussources"
+)
+
+// correlation is a grouping of an endpoint, the observing
+// observer, and a receiver, if any. It's used to unify
+// emitted log record content from evaluated status sources
+// such that no individual source lacks information otherwise
+// only available to others. At this time this means that
+// metrics and component logs will have access to a receiver's
+// observing observer via the Notify event where not available
+// through another means.
+type correlation struct {
+	lastState   endpointState
+	lastUpdated time.Time
+	endpoint    observer.Endpoint
+	receiverID  config.ComponentID
+	observerID  config.ComponentID
+}
+
+// correlationStore provides a centralized interface for up-to-date correlations
+// and receiver attributes as a message passing mechanism by observed components.
+// It manages a reaping loop to prevent stale endpoint buildup over time.
+type correlationStore interface {
+	UpdateEndpoint(endpoint observer.Endpoint, state endpointState, observerID config.ComponentID)
+	GetOrCreate(receiverID config.ComponentID, endpointID observer.EndpointID) correlation
+	Attrs(receiverID config.ComponentID) map[string]string
+	UpdateAttrs(receiverID config.ComponentID, attrs map[string]string)
+	// Start the reaping loop to prevent unnecessary endpoint buildup
+	Start()
+	// Stop the reaping loop
+	Stop()
+}
+
+// store is a collection of mappings used as an instantaneous record of
+// 1. endpoints to their associated receivers->correlations
+// 2. receivers to their endpoint-agnostic Attrs used as a message
+// passing mechanism (currently just for embedded config values).
+// This way the pre-created receiver instances can have attrs
+// before they are created via endpoint.
+type store struct {
+	logger *zap.Logger
+	// ~map[endpointID]map[receiverID]*corr
+	correlations  *sync.Map
+	endpointLocks *keyLock
+	receiverAttrs *sync.Map
+	receiverLocks *keyLock
+	sentinel      chan struct{}
```

Comment:
What do you think about changing the name of this field to `stop`?

Reply:
It's a `chan` so it's kind of both: when the sentinel value is passed to it the reaper loop returns:

```
case <-s.sentinel:
  timer.Stop()
  return
```

I prefer its name to `stop` since that's a verb and this isn't a function.